internal/bcdb/transaction_processor.go

```diff
@@ -266,37 +264,47 @@ func (t *transactionProcessor) SubmitTransaction(tx interface{}, timeout time.Du
 		return nil, err
 	}
 
-	t.Lock()
-	duplicate, err := t.isTxIDDuplicate(txID)
-	if err != nil {
-		t.Unlock()
-		return nil, err
-	}
-	if duplicate {
-		t.Unlock()
+	// We attempt to insert the txID atomically.
+	// If we succeed, then future TX fill fail at this point.
+	// However, if the TX already exists in the block store, then we will fail the subsequent check.
+	// Since a TX will be removed from the pending queue only after it is inserted to the block store,
+	// then it is guaranteed that we won't use the same txID twice.
+	// TODO: add limit on the number of pending sync tx
```

Comment:
Placing a limit will be a problem with sync.Map

Reply:
Outdated. Reverted back to map.